### PR TITLE
Make testing spack commands simpler

### PR DESCRIPTION
**Pull request by tgamblin**
This makes it easier to test command-line commands from within Spack's testing framework.  It adds a `SpackCommand` class so you can do this:

```python
from spack.main import SpackCommand
uninstall = SpackCommand('uninstall')
out, err = uninstall('-y', '-a', '--dependents', 'mpich')
```

No more creating parsers and all that from within command tests.  Just run the command like a user would and test that.  The commands return their error and their output so you can parse that, as well.  I've redone the command tests to use this instead of all the cruft that was there before.  I also removed a bunch of unnecessary mockery from the `spack install` test, and simplified the mocking of fetches a bit.

I hope this will make it easier to get coverage in some of the commands where we've been missing it.

- [x] Add `SpackCommand` to make testing commands easier
- [x] Rework tests of commands to use `SpackCommand`
- [x] Simplify mocking in the `cmd/install` test and in general
- [x] Move the `mock_archive` fixture up to the `conftest.py` level so all tests can use it
- [x] Make a `mock_fetch` fixture usable by all tests

### CODE REVIEW

File: lib/spack/spack/test/cmd/url.py [hunk 1]
```diff
@@ -77,51 +72,64 @@ def test_version_parsed_correctly():
...
     # No version in URL
-    args = parser.parse_args(['parse', 'http://www.netlib.org/voronoi/triangle.zip'])
-    url(parser, args)
+    with pytest.raises(UndetectableVersionError):
+        url('parse', 'http://www.netlib.org/voronoi/triangle.zip')
 
 
-def test_url_list(parser):
-    args = parser.parse_args(['list'])
-    total_urls = url_list(args)
```
Comment (adamjstewart) [inline lib/spack/spack/test/cmd/url.py]: Many of these subcommands were designed to return the total number of URLs for testing purposes. Is that no longer possible?

File: lib/spack/spack/test/cmd/url.py [hunk 2]
```diff
@@ -77,51 +72,64 @@ def test_version_parsed_correctly():
...
-def test_url_parse(parser):
-    args = parser.parse_args(['parse', 'http://zlib.net/fossils/zlib-1.2.10.tar.gz'])
-    url(parser, args)
+def test_url_parse():
+    url('parse', 'http://zlib.net/fossils/zlib-1.2.10.tar.gz')
 
 
-@pytest.mark.xfail
-def test_url_parse_xfail(parser):
+def test_url_with_no_version_failes():
```
Comment (adamjstewart) [inline lib/spack/spack/test/cmd/url.py]: typo: failes -> fails

File: lib/spack/spack/test/cmd/url.py (hunk 1 above)
Comment (tgamblin) [inline lib/spack/spack/test/cmd/url.py]: It's definitely still possible.  I still call some directly below.  Do you want these changed back?  Or should we do both?  It's hard to say which level of testing (command or sub-function) is the right one.  It's d be nice to test both as some of the corner cases in Python 3 porting and other spots arose in commands.

Comment (adamjstewart) [inline lib/spack/spack/test/cmd/url.py]: Hmm, I guess it's not necessary anymore. In fact, you could even modify `spack url` to not return anything anymore. Back when I wrote it, I don't think it was possible to get the output like you're doing.

Comment (tgamblin) [inline lib/spack/spack/test/cmd/url.py]: Let's just keep it -- if we don't need the return val that's fine for now.  The return value of the one that returns dicts is still useful and I kept it in b/c it wasn't worth parsing.  There's a balance between what's easy to test one way or the other.

File: lib/spack/spack/test/cmd/gpg.py [hunk 3]
```diff
@@ -48,134 +60,103 @@ def has_gnupg2():
         return False
 
 
-@pytest.mark.usefixtures('testing_gpg_directory')
```
Comment (mathstuf) [inline lib/spack/spack/test/cmd/gpg.py]: Did I just use pytest wrong? Or is this done instead by pytest magic™?

Comment (alalazo) [inline lib/spack/spack/test/cmd/gpg.py]: @mathstuf It's done by pytest magic by what they call the *funcargs* dependency injection mechanism. The function that you register as a fixture with the decorator:
```python
@pytest.fixture()
def testing_gpg_directory(...):
    ...
```
acts as a factory for a fixture object. The dependency injection mechanism kicks in when in a test function you declare a function argument with the same name as a registered fixture:
```python
def test_gpg(testing_gpg_directory):
    ...
```

What you have done is fine either, the only difference being that you don't have direct access to the created object.

File: lib/spack/spack/test/cmd/gpg.py [hunk 4]
```diff
@@ -48,134 +60,103 @@ def has_gnupg2():
...
+@pytest.mark.xfail  # TODO: fix failing tests.
```
Comment (alalazo) [inline lib/spack/spack/test/cmd/gpg.py]: Why this test has been marked as `xfail`?

File: lib/spack/spack/test/cmd/gpg.py [hunk 5]
```diff
@@ -48,134 +60,103 @@ def has_gnupg2():
...
     # Verify a file with an empty keyring.
-    args = parser.parse_args(['verify', os.path.join(
-        spack.mock_gpg_data_path, 'content.txt')])
     with pytest.raises(ProcessError):
-        gpg.gpg(parser, args)
+        gpg('verify', os.path.join(spack.mock_gpg_data_path, 'content.txt'))
 
     # Import the default key.
-    args = parser.parse_args(['init'])
-    args.import_dir = spack.mock_gpg_keys_path
```
Comment (mathstuf) [inline lib/spack/spack/test/cmd/gpg.py]: @gartung @tgamblin This removal broke the tests from using its own testing keys.
